knitr/convert-odes/convert_odes.Rmd

```diff
@@ -0,0 +1,398 @@
+are packed into the `real[] theta` argument. `kappa` isn't actually a model
+parameter so it is not clear why it is packed in with the other parameters,
+but it is. Promoting `kappa` to a parameter causes there to be more states
+in the extended ODE sensitivity system (used to get gradients of the ODE
+with respect to inputs). The ODE system function looks like:
+
+```{stan, output.var = "", eval = FALSE}
+functions {
+  // theta[1] = beta, water contact rate
+  // theta[2] = kappa, C_{50}
+  // theta[3] = gamma, recovery rate
+  // theta[4] = xi, bacteria production rate
+  // theta[5] = delta, bacteria removal rate
+  real[] simple_SIR(real t,
+                    real[] y,
+                    real[] theta,
+                    real[] x_r,
+                    int[] x_i) {
+    real dydt[4];
+
+    dydt[1] = - theta[1] * y[4] / (y[4] + theta[2]) * y[1];
+    dydt[2] = theta[1] * y[4] / (y[4] + theta[2]) * y[1] - theta[3] * y[2];
+    dydt[3] = theta[3] * y[2];
+    dydt[4] = theta[4] * y[2] - theta[5] * y[4];
+
+    return dydt;
+  }
+}
+```
+
+For comparison, with the new interface the ODE system function can be
+rewritten to explicitly name all the parameters. No separation of data
+and parameters is necessary either -- the solver will not add more
+equations for arguments that are defined in the `data` and
+`transformed data` blocks. The state variables in the new model are also
+represented by `vector` variables instead of `real[]` variables.
+The new ODE system function is:
+
+```{stan, output.var = "", eval = FALSE}
+functions {
+  vector simple_SIR(real t,
+                    vector y,
+                    real beta,    // water contact rate
+                    real kappa,   // C_{50}
+                    real gamma,   // recovery rate
+                    real xi,      // bacteria production rate
+                    real delta) { // bacteria removal rate
+    vector[4] dydt;
+
+    dydt[1] = -beta * y[4] / (y[4] + kappa) * y[1];
+    dydt[2] = beta * y[4] / (y[4] + kappa) * y[1] - gamma * y[2];
+    dydt[3] = gamma * y[2];
+    dydt[4] = xi * y[2] - delta * y[4];
+
+    return dydt;
+  }
+}
+```
+
+### Calling the ODE Solver
+
+In the old ODE interface, the parameters are all packed into a `real[]` array
+before calling the ODE solver:
+
+```{stan, output.var = "", eval = FALSE}
+transformed parameters {
+  real<lower=0> y[N_t, 4];
+  {
+    real theta[5] = {beta, kappa, gamma, xi, delta};
+    y = integrate_ode_rk45(simple_SIR, y0, t0, t, theta, x_r, x_i);
+  }
+}
+```
+
+In the new ODE interface each of the arguments is appended on to the ODE
+solver function call. The RK45 ODE solver with default tolerances is called
+`ode_rk45`. Because the states are handled as `vector` variables, the solver
+output is an array of vectors (`vector[]`).
+
+```{stan, output.var = "", eval = FALSE}
+transformed parameters {
+  vector<lower=0>[4] y[N_t];
+  {
+    y = ode_rk45(simple_SIR, y0, t0, t, beta, kappa, gamma, xi, delta);
```

Comment:
we don't need the inner block any more. So just write

`vector<lower=0>[4] y[N_t] = ode_rk45(simple_SIR, y0, t0, t, beta, kappa, gamma, xi, delta);`

and maybe emphasise that no more temporary is needed for packing purposes.